### ISSUE TITLE
Support for Launch Template - Instance Interruption Behaviour

### ISSUE THREAD

**Issue by gjtempleton** [kubernetes/kops#8809]
**1. Describe IN DETAIL the feature/behavior/change you would like to see.**

Add support for launch template `Instance Termination Behaviour`.

We currently have stubs within both the CF and Terraform targets: 

https://github.com/kubernetes/kops/blob/8df55b8571c1cd6dc5e6f7fba56175bf0945fb74/upup/pkg/fi/cloudup/awstasks/launchtemplate_target_cloudformation.go#L66
https://github.com/kubernetes/kops/blob/8df55b8571c1cd6dc5e6f7fba56175bf0945fb74/upup/pkg/fi/cloudup/awstasks/launchtemplate_target_terraform.go#L65

However there's no way to actually leverage this yet.

This would allow users to override the default `terminate` behaviour of spot instance interruption with either `hibernate` or `stop`.

I'd propose also updating the underlying field from `InstancesInterruptionBehavior` to `InstanceInterruptionBehavior` to match the underlying SDK naming.

**Comment by tomesm** [comment]
@gjtempleton Hi Im a kubernetes (and kops) contributor novice but I'd like to try to solve this. to understand more of the context of this issue can you please tell me if my understanding is correct?

I believe (from very high level) I need to update the struct `InstanceGroups` and `LaunchTemplate`, check If we receive the `InstancesInterruptionBehavior` parameters and if yes properly add them to the LaunchTemplate. I think that this task will be similar (not the same though) like #8802 .

In other words we alrady have the stubs within the CF and Terraform and we need to use tham (make operational), right? 

(I can provide better "design" tomorrow, I am just trying to get familiar with the codebase and with what's going on). I do not want to open a PR before understanding the request.

**Comment by gjtempleton** [comment]
Hey @tomesm 

Welcome! Good to see you getting stuck in!

Sorry for the delay in getting back to you.

You seem to have grabbed the essentials already and that existing PR is indeed a good guide to what you'll need to do to enable the behaviour.

As you'll see with it it'll be worth adding some validation in at the same time given AWS only supports 3 values here.

Happy to give any work you do on this a review.

**Comment by tomesm** [comment]
Hi @gjtempleton, thanks for reply. Ok, gimme few days and I will prepare a PR linked to this issue. A review and feedbeck will be surely needed :)

**Comment by tomesm** [comment]
Hi @gjtempleton I have added
`InstanceInterruptionBehavior *string `json:"instanceInterruptionBehavior,omitempty"``
into InstanceGroupSpec struct.

How do I figure out if the value is simple string or a pointer to a string? Im not sure about that.

**Comment by gjtempleton** [comment]
Are you meaning whether you declare it in the struct as 

`InstanceInterruptionBehavior *string `json:"instanceInterruptionBehavior,omitempty"`

or

`InstanceInterruptionBehavior string `json:"instanceInterruptionBehavior,omitempty`?

Since pointers can be nil and this is an optional field, it'll allow you to do any required validation on whether the value's `nil`. In addition the existing low level fields, including at the AWS API utilise a string pointer, so if you went with a string up here, you'd end up having to pass it as a pointer at some point anyway.

**Comment by tomesm** [comment]
Hi @gjtempleton I have made the changes in my local repo. But I am getting errors when run `make verify `: 

```
To fix: run 'make crds'
make: *** [verify-generate] Error 1
```

This error tells me nothing in detail about an error. I have run make crds and files got generated but no hint about why the error happened. Are there any log files hidden? I dunno how to debug this.

Also make test produces FAIL test on on terraform and cloudformation (this is because my modifications)

`I0429 10:07:30.257994   68167 target.go:221] Terraform output is in /tmp/kops-render-372929841`

Sadly, there are no `kops-render*` files in my `/tmp` folder (I am on OSX). So I am also not able to debug this. Perhaps switch to linux? But I suppose it does not matter, it should produce output in unix `/tmp `folder. Any hint here? 

Or should I go ahead with PR so that it can be reviewd and we will see the failed tests results here on github?

**Comment by rifelpet** [comment]
Try running ./hack/update-expected.sh this will regenerate the expected outputs for many of the tests now that a new field has been added. And yea feel free to open a PR and we can address anything else that needs to happen :)

**Comment by gjtempleton** [comment]
/close

Resolved by #9024
